src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/SuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            var builder = ImmutableDictionary.CreateBuilder<SyntaxNode, AttributeData>();
+            AddAttributes(_compilation.Assembly, _suppressMessageAttributeType, builder);
+            foreach (var module in _compilation.Assembly.Modules)
+            {
+                AddAttributes(module, _suppressMessageAttributeType, builder);
+            }
+
+            return builder.ToImmutable();
+
+            // Local functions.
+            static void AddAttributes(ISymbol symbol, INamedTypeSymbol suppressMessageAttributeType, ImmutableDictionary<SyntaxNode, AttributeData>.Builder builder)
+            {
+                foreach (var attribute in symbol.GetAttributes())
+                {
+                    if (suppressMessageAttributeType.Equals(attribute.AttributeClass) &&
+                        attribute.ApplicationSyntaxReference?.GetSyntax() is SyntaxNode node)
+                    {
+                        builder.Add(node, attribute);
+                    }
+                }
+            }
+        }
+
+        public bool IsGlobalSuppressMessageAttribute(SyntaxNode attributeSyntax, [NotNullWhen(returnValue: true)] out AttributeData? attribute)
+            => _lazySuppressMessageAttributesBySyntax.Value.TryGetValue(attributeSyntax, out attribute);
+
+        public bool HasInvalidScope(AttributeData attribute, out TargetScope targetScope)
+        {
+            if (!TryGetNamedArgument(attribute, SuppressMessageScope, out var scopeString) ||
+                string.IsNullOrEmpty(scopeString))
+            {
+                // Missing/Null/Empty scope values are treated equivalent to a compilation wide suppression.
+                targetScope = TargetScope.Module;
+            }
+            else if (!s_targetScopesMap.TryGetValue(scopeString!, out targetScope))
+            {
+                targetScope = TargetScope.None;
+                return true;
+            }
+
+            return false;
+        }
+
+        public bool HasInvalidOrMissingTarget(AttributeData attribute, TargetScope targetScope)
+        {
+            if (targetScope == TargetScope.Resource)
+            {
+                // Legacy scope which we do not handle.
+                return false;
+            }
+
+            if (!TryGetNamedArgument(attribute, SuppressMessageTarget, out var targetSymbolString))
+            {
+                targetSymbolString = null;
+            }
+
+            if (targetScope == TargetScope.Module)
+            {
+                // Compilation wide suppression with a non-null target is considered invalid.
+                return targetSymbolString != null;
+            }
+
+            var resolvedSymbols = s_listPool.Allocate();
+            try
+            {
+                var resolver = new TargetSymbolResolver(_compilation, targetScope, targetSymbolString);
+                resolvedSymbols.Clear();
+                resolver.Resolve(resolvedSymbols);
```

Comment:
💡 We should update `Resolve` to allow a null argument, and return the number of matched items. The pool approach is not efficient at scale. See #43715 for a similar scenario.

Reply:
Thanks, I will file a separate tracking issue for this.

Reply:
https://github.com/dotnet/roslyn/issues/44343